data/src/main/java/com/linkedin/data/schema/annotation/PegasusExtensionSchemaAnnotationHandler.java

```diff
@@ -0,0 +1,147 @@
+          {
+            appendCompatibilityMessage(result, CompatibilityMessage.Impact.ANNOTATION_INCOMPATIBLE_CHANGE,
+                "Updating extension annotation field: \"%s\" value is considering as a backward incompatible change.",
+                key, currContext.getPathSpecToSchema());
+          }
+          currAnnotations.remove(key);
+        }
+        else
+        {
+          // An existing annotation field is removed.
+          appendCompatibilityMessage(result, CompatibilityMessage.Impact.ANNOTATION_INCOMPATIBLE_CHANGE,
+              "Removing extension annotation field: \"%s\" is considering as a backward incompatible change.",
+              key, currContext.getPathSpecToSchema());
+        }
+      });
+
+      currAnnotations.forEach((key, value) ->
+      {
+        // Adding an extension annotation field.
+        appendCompatibilityMessage(result, CompatibilityMessage.Impact.ANNOTATION_INCOMPATIBLE_CHANGE,
+            "Adding extension annotation field: \"%s\" is a backward incompatible change.",
+            key, currContext.getPathSpecToSchema());
+      });
+    }
+    else if (prevResolvedProperties.containsKey(EXTENSION_ANNOTATION_NAMESPACE))
+    {
+      // Only previous schema has extension annotation, it means the extension annotation is removed in the current schema.
+      if (currContext.getPathSpecToSchema() != null)
+      {
+        appendCompatibilityMessage(result, CompatibilityMessage.Impact.ANNOTATION_INCOMPATIBLE_CHANGE,
+            "Removing extension annotation is a backward incompatible change.",
+            null, prevContext.getPathSpecToSchema());
+      }
+      else
+      {
+        // an existing field with extension annotation is removed
+        appendCompatibilityMessage(result, CompatibilityMessage.Impact.ANNOTATION_INCOMPATIBLE_CHANGE,
+            "Removing field: \"%s\" with extension annotation is a backward incompatible change.",
+            prevContext.getSchemaField().getName(), prevContext.getPathSpecToSchema());
+      }
+    }
+    else
+    {
+      if (prevContext.getPathSpecToSchema() != null)
+      {
+        appendCompatibilityMessage(result, CompatibilityMessage.Impact.ANNOTATION_INCOMPATIBLE_CHANGE,
+            "Adding extension annotation on an existing field: \"%s\" is backward incompatible change",
+            prevContext.getSchemaField().getName() , currContext.getPathSpecToSchema());
+      }
+      else
+      {
+        // Adding a new injected field with extension annotation.
+        appendCompatibilityMessage(result, CompatibilityMessage.Impact.ANNOTATION_COMPATIBLE_CHANGE,
+            "Adding extension annotation on new field: \"%s\" is backward compatible change", currContext.getSchemaField().getName() , currContext.getPathSpecToSchema());
+      }
+    }
+    return result;
+  }
+
+  private void appendCompatibilityMessage(AnnotationCompatibilityResult result, CompatibilityMessage.Impact impact, String message, String context, PathSpec pathSpec)
```

Comment:
The above logic only has 1 `ANNOTATION_COMPATIBLE_CHANGE` impact, all the other impact are incompatible.
I wonder how is it better if split into two results (or just one result if you dont care comptaible message, btw why do we need compatible message?)
If we split to two results, and you can add all incompatible msg to a dedicated result. That will make these callers that are saving msg not need to pass the `CompatibilityMessage.Impact.ANNOTATION_INCOMPATIBLE_CHANGE` and less repeating ourselves.

Reply:
We only have ONE result here. result contains a message list which contains all messages no matter it is compatible or incompatible. We will not throw the exception for compatible change, but we do need such information to generate compatibility report. That will be the same user experience as schema compatibility check.